lib/charms/tempo_k8s/v1/charm_tracing.py

```diff
@@ -560,3 +660,111 @@ def trace(obj: Union[Type, Callable]):
             raise UntraceableObjectError(
                 f"cannot create span from {type(obj)}; instrument {obj} manually."
             )
+
+
+def _dict_to_status(value: Dict) -> _EntityStatus:
+    return _EntityStatus(**value)
+
+
+def _dict_to_model(value: Dict) -> Model:
+    return Model(**value)
+
+
+def _dict_to_relation(value: Dict) -> "AnyRelation":
+    relation_type = value.pop("relation_type")
+    if relation_type == "Relation":
+        return Relation(**value)
+    if relation_type == "PeerRelation":
+        return PeerRelation(**value)
+    if relation_type == "SubordinateRelation":
+        return SubordinateRelation(**value)
+    raise TypeError(value)
+
+
+def _dict_to_address(value: Dict) -> Address:
+    return Address(**value)
+
+
+def _dict_to_bindaddress(value: Dict) -> BindAddress:
+    if addrs := value.get("addresses"):
+        value["addresses"] = [_dict_to_address(addr) for addr in addrs]
+    return BindAddress(**value)
+
+
+def _dict_to_network(value: Dict) -> Network:
+    if addrs := value.get("bind_addresses"):
+        value["bind_addresses"] = [_dict_to_bindaddress(addr) for addr in addrs]
+    return Network(**value)
+
+
+def _dict_to_container(value: Dict) -> Container:
+    if layers := value.get("layers"):
+        value["layers"] = {l_name: pebble.Layer(l_raw) for l_name, l_raw in layers.items()}
+    return Container(**value)
+
+
+def _dict_to_opened_port(value: Dict) -> Port:
+    return Port(**value)
+
+
+def _dict_to_secret(value: Dict) -> Secret:
+    if rotate := value.get("rotate"):
+        value["rotate"] = SecretRotate(rotate)
+    if expire := value.get("expire"):
+        value["expire"] = datetime.fromisoformat(expire)
+    return Secret(**value)
+
+
+def _dict_to_stored_state(value: Dict) -> StoredState:
+    return StoredState(**value)
+
+
+def _dict_to_deferred(value: Dict) -> DeferredEvent:
+    return DeferredEvent(**value)
+
+
+def _dict_to_storage(value: Dict) -> Storage:
+    return Storage(**value)
+
+
+def dict_to_state(state_json: Dict) -> State:  # noqa: C901
```

Comment:
I do see why you want to put conversions both ways (both `state -> dict` and `dict -> state`) close to the place where it's used but I wonder if that's part of `charm_tracing` area of responsibility.

Especially this method (in the dict -> state direction), together with the helper methods above, is currently essentially used within the tests? I'm thinking if it might be a better fit within the `snapshot` lib. Or a converter class that _could_ be outside of charm libs altogether, maybe closer to scenario? I can imagine this being useful in other contexts than `charm_tracing` as well. Please correct me if I missed something or don't yet see where that's heading.

Reply:
you're right, it's ugly. 

IMHO: 
- the snapshot lib is responsible for generating a State object
- the charm (user code) shouldn't be responsible for serializing it to something that the charm_tracing lib can consume. 
- charm_tracing needs 

So either we make the get_state function return an opaque serialized object that the charm user can't inspect (might be fine if we want to prevent misuse, but misuse is in this case also a lot of fun) and simply has to pass through to the charm_tracing lib

Or, perhaps that's better, we keep it outside of the charm concern altogether. I think I have an idea.

Reply:
(updated)